gin/config.py

```diff
@@ -270,13 +270,13 @@ class DecoratedClass(fn_or_cls):
   elif subclass:
     # this ensures picklability, as we produce the original class instance,
     # the implicit inheritance is skipped
-    def hijacking_new(klass, *args, **kwargs):
+    def _new(klass, *args, **kwargs):
       if klass is DecoratedClass:
         klass = fn_or_cls
       obj = fn_or_cls.__new__(klass, *args, **kwargs)
       decorated_fn(obj, *args, **kwargs)
```

Comment:
Thanks for the change! Initially I thought this would work, but after testing it internally a couple of issues came up. Specifically:

1) Calling decorated_fn here can break scoping. In order to apply config scopes to references (as in "@some_scope_name/ConfigurableClass()"), the configurable version of the function gets wrapped by yet another function that applies the scope. This happens when constructing a ConfigurableReference instance here, and reuses all the same decoration mechanisms (including subclassing for classes) as regular gin.configurable decoration does:

  https://github.com/google/gin-config/blob/484d03ab6f52c05455adb06541e9d8d0189d6967/gin/config.py#L301

  So for scoped configurable references, there are two layers of wrapping, one to apply the scope, and one to supply the configured parameters. Unfortunately, for that outer layer of wrapping, calling __new__ first before calling decorated_fn means that __new__ is called before the config_scope is opened (since that happens inside decorated_fn). That inner __new__ call then in turn calls its own decorated_fn (which is responsible for supplying parameters), but since it's not in the right scope the wrong parameters can be supplied. Hopefully that's somewhat clear...

2) The second issue is that this breaks for classes that use metaclasses, when the metaclass overrides __call__ and calls __init__ internally. One fairly broadly used example of this is in Sonnet v2's ModuleMetaclass here

 
 https://github.com/deepmind/sonnet/blob/5a465696383f967d5bffb6599347d9e6c15cef4b/sonnet/src/base.py#L126

  Here's a simple test case that fails for that reason (requires Sonnet):

  ```
import sonnet.v2 as snt

configurable_linear = gin.external_configurable(snt.Linear)

...

def testConfigurableSonnetModule(self):
    config_str = """
      Linear.output_size = 100
    """
    gin.parse_config(config_str)
    linear = configurable_linear()
    self.assertEqual(linear.output_size, 100)
```

Unfortunately I think those issues are hard to fix using the current approach... I haven't yet had a chance to think through alternative implementations in any depth, so not sure what other approaches might be feasible. Do you have any thoughts on another way to address this? (One approach that I'd like to get to is to replace a lot of this quite complicated ad-hoc decoration logic with wrapt, which would certainly change a lot of things implementation wise, but it's not clear to me it would fix the pickling issue specifically.)

Reply:
Thanks for the review, I'll take a look into the points you mentioned.

The first one I didn't anticipate - the tests didn't fail, so I supposed it's all okay. I'll try to address that with a test case in the meantime.

The second seems more like a trade-off, as this is much more of what's doable with the configurable class. I know the priority is to keep the old stuff working, but an expectation to pickle the result of computations is quite standard I believe. I will think about keeping both for sure, but it's not obvious that there's a space for that.

Basically to ensure picklability, a class definition must be "static" (as static as for Python it could be) - it must be importable, it cannot be defined within a function. I'll review [the more complicated approach](https://stackoverflow.com/a/4648949/3067956) I found, but I will need to re-think how that's doable - an importable class must be accessible at the module scope anyway.

Reply:
Yeah it's unfortunate the first point isn't covered by the current tests... scoping is generally well tested but apparently scopes + references to external configurables seems to have been an overlooked case. A contribution there would be very welcome!

For the second point, since Sonnet is kind of an important use case I think it's necessary to maintain compatibility there. That said, an approach using __reduce(_ex)__ seems like it could be a good direction, and I'd think is more likely to be compatible with the current setup. Let me know if you look into it more and can tell if it would be feasible. Thanks again for working on this!

Reply:
@dhr, based on [this](https://stackoverflow.com/a/6966942/3067956) and [this](https://stackoverflow.com/a/6798042/3067956) I think that the proper way to solve the second case is to actually wrap the `klass.__metaclass__.__call__` function directly, as that's what I was (indirectly) trying to achieve with my change. It would be a generalization of the current version of this fix for any metaclass and would probably address the first issue as well.

Reply:
Also, right now you don't address the metaclasses in any specific way - you patch `__new__` or `__init__` only ([link](https://github.com/google/gin-config/blob/master/gin/config.py#L192-L198)). I'll prepare a test that replicates the Sonnet scenario you pointed out.

Reply:
I've replicated the Sonnet issue in tests.

The solution with `klass.__metaclass__.__call__` will however slightly change the legacy behavior. `klass.__metaclass__.__call__` parameters will get patched, instead of `klass.__init__`.

I've found out that there was a difference between parameters going to `klass.__new__` and `klass.__init__`, which actually should be an issue. If your `klass.__new__` was supposed to return another class (let's say `int`), but had `klass.__init__` defined, it would return `int` only when parameters are passed in the code, not in the config. Here's test case describing the current behavior:

```python
@config.configurable
class ConfigurableClassWithNew(object):
  """A configurable class with __new__."""

  def __new__(cls, return_none=False):
    if return_none: # this will test __new__ and __init__ separation
      return None
    return super(ConfigurableClassWithNew, cls).__new__(cls)

  def __init__(self, return_none=False):
    pass

...

  def testConfigurableClassWithNewCanReturnOtherType(self):
    config_str = """
      ConfigurableClassWithNew.return_none = True
    """
    config.parse_config(config_str)
    instance = ConfigurableClassWithNew()
    self.assertIsNotNone(instance)

    config_str = ""
    config.parse_config(config_str)
    instance = ConfigurableClassWithNew(return_none=True)
    self.assertIsNone(instance)
```

After patching `klass.__metaclass__.__call__` in both cases we will have `None` (more what I'd expect). @dhr, @sguada will that change break antyhing for you?

Reply:
Works with `__reduce__` 🙂